Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs

```diff
@@ -853,6 +854,98 @@ protected internal uint stateOffsetRelativeToDeviceRoot
         internal PrimitiveValue m_MinValue;
         internal PrimitiveValue m_MaxValue;
 
+        internal FourCC m_OptimizedControlDataType;
+
+        /// <summary>
+        /// For some controls it's possible to safely read/write state memory directly
+        /// instead of calling ReadUnprocessedValueFromState/WriteValueIntoState.
+        /// This value will represent a type that should be used for reading/writing directly.
+        /// Or be InputStateBlock.kFormatInvalid if no casting is possible.
+        /// </summary>
+        /// <remarks>
+        /// For example AxisControl might be a "float" in state memory, and if no processing applied during reading (e.g. no invert/scale/etc),
+        /// then it could be read as float in memory directly without calling ReadUnprocessedValueFromState.
+        /// If then Vector3Control has 3 AxisControls as consecutive floats in memory,
+        /// then we can cast Vector3Control state memory directly to Vector3 without calling ReadUnprocessedValueFromState on x/y/z axes.
+        ///
+        /// Value is computed by automatically calling <see cref="InputControl.CalculateOptimizedControlDataType"/> when system believes control changes setup configuration.
+        /// Value can be updated manually by calling <see cref="InputControl.ApplyParameterChanges"/>.
+        /// </remarks>
+        public FourCC optimizedControlDataType => m_OptimizedControlDataType;
+
+        /// <summary>
+        /// Calculate and return a type that can represent control value in memory directly.
+        /// Value then is cached in <see cref="InputControl.optimizedControlDataType"/>.
+        /// Function shouldn't be called manually.
+        /// </summary>
+        protected virtual FourCC CalculateOptimizedControlDataType()
+        {
+            return InputStateBlock.kFormatInvalid;
+        }
+
+        /// <summary>
+        /// Apply built-in parameters changes (e.g. <see cref="AxisControl.invert"/>, others) and recompute <see cref="InputControl.optimizedControlDataType"/> for impacted controls.
+        /// </summary>
+        /// <remarks>
+        /// </remarks>
+        public void ApplyParameterChanges()
+        {
+            // First we go through all children of our own hierarchy
+            SetOptimizedControlDataTypeRecursively();
```

Comment:
Or just computing status for current control is sufficient?

Reply:
Keep as is for now, probably safer that way as `ApplyParameterChanges` on root (device) will force-recompute all the states, so kinda an escape hatch of sorts